cpp/inc/bond/core/detail/omit_default.h

```diff
@@ -4,6 +4,10 @@
 #pragma once
 
 #include <bond/core/config.h>
+#include <bond/core/bond_fwd.h>
+#include <bond/core/maybe.h>
+#include <bond/core/reflection.h>
+#include "odr.h"
```

Comment:
Why do we need `odr.h` and `reflection.h`?

Reply:
Because of this line [here](https://github.com/microsoft/bond/blob/adcecd23e56d3a730f94ff644c4483be5b391435/cpp/inc/bond/core/detail/omit_default.h#L96):
`(void)one_definition<may_omit_fields<Writer>, std::true_type>::value;`

Reply:
Ok, then that will cover the `is_basic_type` only. The other two are from `bond/core/stl_containers.h` and `bond/core/traits.h`.

Reply:
Somehow those two work, but I'll add `stl_containers.h` and `traits.h` as well.

Now, for repro:
Add `#include <bond/core/transforms.h>` as a first line to, say, `bond\cpp\test\core\allocator_test.cpp`, get these errors:
```
bond\cpp\inc\bond\core\detail\omit_default.h(27,1): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
bond\cpp\inc\bond\core\detail\omit_default.h(27,28): error C2988: unrecognizable template declaration/definition
bond\cpp\inc\bond\core\detail\omit_default.h(27,28): error C2143: syntax error: missing ',' before '<'
bond\cpp\inc\bond\core\detail\omit_default.h(29,18): error C2027: use of undefined type 'bond::value'
bond\cpp\inc\bond\core\bond_fwd.h(36): message : see declaration of 'bond::value'
bond\cpp\inc\bond\core\detail\omit_default.h(36,29): error C2065: 'is_basic_type': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(37,1): error C2588: '::!is_string_type': illegal global finalizer
bond\cpp\inc\bond\core\detail\omit_default.h(37,43): error C2988: unrecognizable template declaration/definition
bond\cpp\inc\bond\core\detail\omit_default.h(37,43): error C2143: syntax error: missing ';' before '<'
bond\cpp\inc\bond\core\detail\omit_default.h(37,43): error C2059: syntax error: '<'
bond\cpp\inc\bond\core\detail\omit_default.h(37,48): error C2039: 'value': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(38,47): error C2039: 'value': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(38,61): error C2039: 'type': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(40,1): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(40,1): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(48,27): error C2065: 'is_type_alias': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(48,17): error C2974: 'boost::enable_if': invalid template argument for 'Cond', type expected
Boost.Library\v1.67.0\include\boost/core/enable_if.hpp(41): message : see declaration of 'boost::enable_if'
bond\cpp\inc\bond\core\detail\omit_default.h(48,17): error C2976: 'boost::enable_if': too few template arguments
Boost.Library\v1.67.0\include\boost/core/enable_if.hpp(41): message : see declaration of 'boost::enable_if'
bond\cpp\inc\bond\core\detail\omit_default.h(48,43): error C2988: unrecognizable template declaration/definition
bond\cpp\inc\bond\core\detail\omit_default.h(48,43): error C2059: syntax error: ','
bond\cpp\inc\bond\core\detail\omit_default.h(48,52): error C2039: 'type': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(50,1): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(50,1): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(58,27): error C2065: 'is_string': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(58,37): error C2226: syntax error: unexpected type 'std::remove_const<_Ty>::type'
bond\cpp\inc\bond\core\detail\omit_default.h(58): error C2062: type 'unknown-type' unexpected
bond\cpp\inc\bond\core\detail\omit_default.h(58,82): error C2039: 'type': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(60,1): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(60,1): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(68,27): error C2065: 'is_wstring': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(68,38): error C2226: syntax error: unexpected type 'std::remove_const<_Ty>::type'
bond\cpp\inc\bond\core\detail\omit_default.h(68): error C2062: type 'unknown-type' unexpected
bond\cpp\inc\bond\core\detail\omit_default.h(68,83): error C2039: 'type': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(70,1): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(70,1): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(79,27): error C2065: 'is_container': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(79,17): error C2974: 'boost::enable_if': invalid template argument for 'Cond', type expected
Boost.Library\v1.67.0\include\boost/core/enable_if.hpp(41): message : see declaration of 'boost::enable_if'
bond\cpp\inc\bond\core\detail\omit_default.h(79,17): error C2976: 'boost::enable_if': too few template arguments
Boost.Library\v1.67.0\include\boost/core/enable_if.hpp(41): message : see declaration of 'boost::enable_if'
bond\cpp\inc\bond\core\detail\omit_default.h(79,42): error C2988: unrecognizable template declaration/definition
bond\cpp\inc\bond\core\detail\omit_default.h(79,42): error C2059: syntax error: ','
bond\cpp\inc\bond\core\detail\omit_default.h(79,51): error C2039: 'type': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(81,1): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(81,1): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(89,27): error C2065: 'is_bond_type': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(89,17): error C2974: 'boost::enable_if': invalid template argument for 'Cond', type expected
Boost.Library\v1.67.0\include\boost/core/enable_if.hpp(41): message : see declaration of 'boost::enable_if'
bond\cpp\inc\bond\core\detail\omit_default.h(89,17): error C2976: 'boost::enable_if': too few template arguments
Boost.Library\v1.67.0\include\boost/core/enable_if.hpp(41): message : see declaration of 'boost::enable_if'
bond\cpp\inc\bond\core\detail\omit_default.h(89,42): error C2988: unrecognizable template declaration/definition
bond\cpp\inc\bond\core\detail\omit_default.h(89,42): error C2059: syntax error: ','
bond\cpp\inc\bond\core\detail\omit_default.h(89,51): error C2039: 'type': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(91,1): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(91,1): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(104,11): error C3861: 'one_definition': identifier not found
bond\cpp\inc\bond\core\detail\omit_default.h(104,68): error C2039: 'value': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(109,1): error C2027: use of undefined type 'bond::Metadata'
bond\cpp\inc\bond\core\bond_fwd.h(67): message : see declaration of 'bond::Metadata'
bond\cpp\inc\bond\core\detail\omit_default.h(107,21): error C2027: use of undefined type 'bond::Metadata'
bond\cpp\inc\bond\core\bond_fwd.h(67): message : see declaration of 'bond::Metadata'
bond\cpp\inc\bond\core\detail\omit_default.h(107,33): error C3861: 'Optional': identifier not found
bond\cpp\inc\bond\core\detail\omit_default.h(119,11): error C3861: 'one_definition': identifier not found
bond\cpp\inc\bond\core\detail\omit_default.h(119,69): error C2039: 'value': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(155,13): error C3083: 'mpl': the symbol to the left of a '::' must be a type
bond\cpp\inc\bond\core\detail\omit_default.h(155,18): error C2039: 'void_t': is not a member of 'bond::detail'
bond\cpp\inc\bond\core\detail\omit_default.h(20): message : see declaration of 'bond::detail'
bond\cpp\inc\bond\core\detail\omit_default.h(155,1): error C2065: 'void_t': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(155,25): error C2059: syntax error: 'decltype'
bond\cpp\inc\bond\core\detail\omit_default.h(158,1): error C2332: 'struct': missing tag name
bond\cpp\inc\bond\core\detail\omit_default.h(158,1): error C3306: 'bond::detail::<unnamed-tag>': unnamed class template is not allowed
bond\cpp\inc\bond\core\detail\omit_default.h(158,35): error C2988: unrecognizable template declaration/definition
bond\cpp\inc\bond\core\detail\omit_default.h(158,35): error C2143: syntax error: missing ';' before '>'
bond\cpp\inc\bond\core\detail\omit_default.h(158,35): error C2059: syntax error: '>'
bond\cpp\inc\bond\core\detail\omit_default.h(160,12): error C2059: syntax error: 'function-style cast'
bond\cpp\inc\bond\core\detail\omit_default.h(160,22): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(160,22): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(227,13): error C3083: 'mpl': the symbol to the left of a '::' must be a type
bond\cpp\inc\bond\core\detail\omit_default.h(227,18): error C2039: 'void_t': is not a member of 'bond::detail'
bond\cpp\inc\bond\core\detail\omit_default.h(20): message : see declaration of 'bond::detail'
bond\cpp\inc\bond\core\detail\omit_default.h(227,1): error C2065: 'void_t': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(227,25): error C2059: syntax error: 'decltype'
bond\cpp\inc\bond\core\detail\omit_default.h(227,1): error C2332: 'struct': missing tag name
bond\cpp\inc\bond\core\detail\omit_default.h(227,74): error C2988: unrecognizable template declaration/definition
bond\cpp\inc\bond\core\detail\omit_default.h(227,74): error C2143: syntax error: missing ';' before '>'
bond\cpp\inc\bond\core\detail\omit_default.h(227,74): error C2059: syntax error: '>'
bond\cpp\inc\bond\core\detail\omit_default.h(229,12): error C2059: syntax error: 'function-style cast'
bond\cpp\inc\bond\core\detail\omit_default.h(229,22): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(229,22): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(237,13): error C3083: 'mpl': the symbol to the left of a '::' must be a type
bond\cpp\inc\bond\core\detail\omit_default.h(237,18): error C2039: 'void_t': is not a member of 'bond::detail'
bond\cpp\inc\bond\core\detail\omit_default.h(20): message : see declaration of 'bond::detail'
bond\cpp\inc\bond\core\detail\omit_default.h(237,1): error C2065: 'void_t': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(237,25): error C2059: syntax error: 'decltype'
bond\cpp\inc\bond\core\detail\omit_default.h(237,1): error C2332: 'struct': missing tag name
bond\cpp\inc\bond\core\detail\omit_default.h(237,94): error C2988: unrecognizable template declaration/definition
bond\cpp\inc\bond\core\detail\omit_default.h(237,94): error C2143: syntax error: missing ';' before '>'
bond\cpp\inc\bond\core\detail\omit_default.h(237,94): error C2059: syntax error: '>'
bond\cpp\inc\bond\core\detail\omit_default.h(239,12): error C2059: syntax error: 'function-style cast'
bond\cpp\inc\bond\core\detail\omit_default.h(239,22): error C2143: syntax error: missing ';' before '{'
bond\cpp\inc\bond\core\detail\omit_default.h(239,22): error C2447: '{': missing function header (old-style formal list?)
bond\cpp\inc\bond\core\detail\omit_default.h(300,10): error C2065: 'uses_static_parser': undeclared identifier
bond\cpp\inc\bond\core\detail\omit_default.h(301): message : see reference to class template instantiation 'bond::may_omit_fields<Writer>' being compiled
bond\cpp\inc\bond\core\detail\omit_default.h(300,29): error C2226: syntax error: unexpected type 'T::Reader'
bond\cpp\inc\bond\core\detail\omit_default.h(300,50): error C2039: 'value': is not a member of '`global namespace''
bond\cpp\inc\bond\core\detail\omit_default.h(300,1): error C3770: 'unknown-type': is not a valid base class
bond\cpp\inc\bond\core\detail\omit_default.h(300,1): error C2146: syntax error: missing ',' before identifier 'value'
bond\cpp\inc\bond\core\detail\omit_default.h(301,1): error C2955: 'bond::value': use of class template requires template argument list
bond\cpp\inc\bond\core\bond_fwd.h(36): message : see declaration of 'bond::value'
bond\cpp\inc\bond\core\detail\omit_default.h(301,9): error C2143: syntax error: missing ',' before '||'
bond\cpp\inc\bond\core\detail/debug.h(17,32): error C2039: 'noncopyable': is not a member of 'boost'
Boost.Library\v1.67.0\include\boost/core/enable_if.hpp(24): message : see declaration of 'boost'
bond\cpp\inc\bond\core\detail/debug.h(52): message : see reference to class template instantiation 'bond::detail::OptionalDefault<T>' being compiled
bond\cpp\inc\bond\core\detail/debug.h(33,34): error C2065: 'is_bond_type': undeclared identifier
bond\cpp\inc\bond\core\detail/debug.h(33,47): error C2226: syntax error: unexpected type 'Field::field_type'
bond\cpp\inc\bond\core\detail/debug.h(33): error C2062: type 'unknown-type' unexpected
bond\cpp\inc\bond\core\detail/debug.h(33,5): fatal error C1003: error count exceeds 100; stopping compilation
```